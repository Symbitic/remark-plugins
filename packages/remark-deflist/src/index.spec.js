/* eslint-env jest */
import deflist from "./index";
import html from "rehype-stringify";
import markdown from "remark-parse";
import remark2rehype from "remark-rehype";
import unified from "unified";
import mdx from "@mdx-js/mdx";
import mdxUtil from "mdast-util-mdx";
import syntax from "micromark-extension-mdxjs";
import remarkSuperSub from "../../remark-supersub/src";

const strip = ([str]) => {
  const stripped = str.replace(/\n {6}/g, "\n").replace(/\n {4}$/, "");
  console.log(stripped);
  return stripped;
};

const parse = (str) =>
  unified()
    .use(markdown)
    .use(deflist)
    .use(remark2rehype)
    .use(html)
    .process(str, function (err, file) {
      if (err) throw err;
      console.log(String(file));
    })
    .then((data) => data.toString());

const mdxParse = (str) =>
  mdx(str, {
    remarkPlugins: [
      [
        deflist,
        {
          fromMarkdownOptions: {
            extensions: [syntax()],
            mdastExtensions: [mdxUtil.fromMarkdown],
          },
          toMarkdownOptions: { extensions: [mdxUtil.toMarkdown] },
        },
      ],
      remarkSuperSub,
    ],
  }).then((data) => data.toString());

const tests = [
  [
    "basic definition list",
    strip`
      Term 1

      : Definition 1
    `,
  ],
  [
    "definition list with inline markup",
    strip`
      Term *1*

      : Definition **1**
    `,
  ],
  [
    "document with other elements",
    strip`
      Definition List

      : Definition 1

      This paragraph follows the definition list.
    `,
  ],
  [
    "definition list with continuation",
    strip`
      Term 1

      : Definition
        with continuation.
    `,
  ],
  [
    "definition list with lazy continuation",
    strip`
      Term 1

      : Definition
      with lazy continuation.
    `,
  ],
  [
    "definition list with no space between the term and the descriptions (#7)",
    strip`
      Term **1**
      : Definition **bold** 1
      : Definition 2
    `,
  ],
  [
    "definition list with multiple descriptions (#9)",
    strip`
      Multiple descriptions

      : Description **1**
      : Description 2
    `,
  ],
  [
    "document with several subsequent definition lists (#10)",
    strip`
      Definition List 1

      : Description 1

      Definition List 2

      : Description 1

      Definition List 3

      : Description 1
    `,
  ],
];

const mdxTests = [
  ...tests,
  [
    "definition list with jsx",
    strip`
      Definition List 1

      : Definition <b>{2 + 5}</b>
    `,
  ],
  [
    "definition list with another remark plugin's syntax",
    strip`
      Definition List 1

      : Definition^1^, H~2~0
    `,
  ],
];

describe("remark-deflist", () => {
  describe.skip.each(tests)("with a standard parser", (name, source) => {
    it(`should parse a ${name}`, () => {
      return expect(parse(source)).resolves.toMatchSnapshot();
    });
  });

  describe.each(mdxTests)("with an mdx parser", (name, source) => {
    it(`should parse a ${name}`, () => {
      return expect(mdxParse(source)).resolves.toMatchSnapshot();
    });
  });
});
