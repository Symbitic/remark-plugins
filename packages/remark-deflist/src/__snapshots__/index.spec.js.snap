// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remark-deflist should parse a basic definition list 1`] = `"<dl><dt>Term 1</dt><dd>Definition 1</dd></dl>"`;

exports[`remark-deflist should parse a definition list with continuation 1`] = `
"<dl><dt>Term 1</dt><dd>Definition
with continuation.</dd></dl>"
`;

exports[`remark-deflist should parse a definition list with inline markup 1`] = `"<dl><dt>Term <em>1</em></dt><dd>Definition <strong>1</strong></dd></dl>"`;

exports[`remark-deflist should parse a definition list with lazy continuation 1`] = `
"<dl><dt>Term 1</dt><dd>Definition
with lazy continuation.</dd></dl>"
`;

exports[`remark-deflist should parse a definition list with multiple descriptions (#9) 1`] = `"<dl><dt>Multiple descriptions</dt><dd>Description <strong>1</strong></dd><dd>Description 2</dd></dl>"`;

exports[`remark-deflist should parse a definition list with no space between the term and the descriptions (#7) 1`] = `"<dl><dt>Term <strong>1</strong></dt><dd>Definition <strong>bold</strong> 1</dd><dd>Definition 2</dd></dl>"`;

exports[`remark-deflist should parse a document with other elements 1`] = `
"<dl><dt>Definition List</dt><dd>Definition 1</dd></dl>
<p>This paragraph follows the definition list.</p>"
`;

exports[`remark-deflist should parse a document with several subsequent definition lists (#10) 1`] = `"<dl><dt>Definition List 1</dt><dd>Description 1</dd><dt>Definition List 2</dt><dd>Description 1</dd><dt>Definition List 3</dt><dd>Description 1</dd></dl>"`;
